models/issue_dependency.go

```diff
@@ -0,0 +1,161 @@
+type DependencyType int
+
+// Define Dependency Types
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (exists, circular bool, err error) {
+	sess := x.NewSession()
+	if err := sess.Begin(); err != nil {
+		return false, false, err
+	}
+	defer sess.Close()
+
+	// Check if it aleready exists
+	exists, circular, err = issueDepExists(sess, issue.ID, dep.ID)
+	if err != nil {
+		return
+	}
+
+	// If it not exists, create it, otherwise show an error message
+	if !exists && !circular {
+		newIssueDependency := &IssueDependency{
+			UserID:       user.ID,
+			IssueID:      issue.ID,
+			DependencyID: dep.ID,
+		}
+
+		if _, err := sess.Insert(newIssueDependency); err != nil {
+			return exists, circular, err
+		}
+
+		// Add comment referencing the new dependency
+		if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
+			return
+		}
+
+		// Create a new comment for the dependent issue
+		if _, err = createIssueDependencyComment(sess, user, dep, issue, true); err != nil {
+			return
+		}
+	}
+	return exists, circular, sess.Commit()
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType DependencyType) (err error) {
+	sess := x.NewSession()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+	defer sess.Close()
+
+	// Check if it exists
+	var exists bool
+	switch depType {
+	case DependencyTypeBlockedBy:
+		exists, _, err = issueDepExists(sess, issue.ID, dep.ID)
+	case DependencyTypeBlocking:
+		exists, _, err = issueDepExists(sess, dep.ID, issue.ID)
+	default:
+		return
+	}
+	if err != nil {
+		return err
+	}
+
+	// If it exists, remove it, otherwise show an error message
+	if exists {
+
+		var issueDepToDelete IssueDependency
+
+		switch depType {
+		case DependencyTypeBlockedBy:
+			issueDepToDelete = IssueDependency{IssueID: issue.ID, DependencyID: dep.ID}
+		case DependencyTypeBlocking:
+			issueDepToDelete = IssueDependency{IssueID: dep.ID, DependencyID: issue.ID}
+		default:
+			return
+		}
+
+		if _, err := sess.Delete(&issueDepToDelete); err != nil {
+			return err
+		}
+
+		// Add comment referencing the removed dependency
+		if _, err = createIssueDependencyComment(sess, user, issue, dep, false); err != nil {
+			return err
+		}
+
+		// Create a new comment for the dependent issue
+		if _, err = createIssueDependencyComment(sess, user, dep, issue, false); err != nil {
+			return err
+		}
+	}
+	return sess.Commit()
+}
+
+// Check if the dependency already exists
+func issueDepExists(e Engine, issueID int64, depID int64) (exists, circular bool, err error) {
+
+	// Check if the dependency exists
+	exists, err = e.Where("(issue_id = ? AND dependency_id = ?)", issueID, depID).Exist(&IssueDependency{})
+
+	// If not, check for circular dependencies
```

Comment:
What about cycles of length 3 or more? That is, issue A depends on issue B, which depends on issue C, which depends on issue A.

Reply:
If we start to check this, we would also need to check everything of length `n` and I don't think this could be done with reasonable performance, which is why I'm only checking 2 issues.